### ISSUE TITLE
error when adding nvbench as subproject

### ISSUE THREAD

**Issue by bsarden** [mesonbuild/meson#10244]
Hello,

I'm just getting started with meson and would like to add [nvbench](https://github.com/NVIDIA/nvbench) as a subproject in my meson project using the `cmake` module. Following the examples in the documentation, I cloned the `nvbench` repo under `subprojects` and added the following lines to my top-level meson.build file.

```python
cmake = import('cmake')
opt_var = cmake.subproject_options()
opt_var.add_cmake_defines({'CMAKE_CUDA_ARCHITECTURE': 70})
nvbench_proj = cmake.subproject('nvbench', options: opt_var)
```

However, when running `meson setup`, we seem to get through the cmake configure process just fine and fail with the following error:

```txt
subprojects/nvbench/meson.build:0:0: ERROR: File nvbench does not exist
```

I can see the generated meson.build file under `builddir/subprojects/nvbench`, but I'm not sure what the next steps are from there. Any help would be appreciated!

OS: Ubuntu 20.04
Meson version: 0.62.0
cmake version: 3.22.3

**Comment by anadon** [comment]
I'm new here too, but I'll see what I can do.  I may just end up being a rubber duck.

Is `nvbench` an installed package, and does `cmake` or `pkg-config` pick it up?  You're using some `cmake` functionality that I'm not familiar with; can you link me to it?

**Comment by bsarden** [comment]
Hey @anadon, thanks for replying!

I'm currently not installing the package, I've just been following the [cmake subprojects](https://mesonbuild.com/CMake-module.html#cmake-subprojects) documentation in Meson, where it has us cloning the project under the `subprojects` directory I believe. As a result, my setup looks something like this:

```txt
<repo-head>
├── builddir
│   ├── meson-logs
│   ├── nvbench
├── subprojects
│   ├── nvbench
│   ├── exec
│   ├── ci
│   ├── scripts
│   ├── docs
│   ├── build
│   ├── cmake
│   ├── examples
│   ├── .git
│   └── testing
│   ├── packagefiles
```
As mentioned above, we seem to get through the configure phase but then the output meson.build file is complaining about files not existing. I've also tried using the provided [cmake2meson.py](https://github.com/mesonbuild/meson/blob/master/tools/cmake2meson.py) script to do the conversion, but that doesn't seem to output anything usable (in fact, that results in syntax errors due to some transmog artifacts I believe).

Thanks again for the help, and let me know if you have any more troubles reproducing :)

**Comment by eli-schwartz** [comment]
I get a different error:

```
| CMake Error in nvbench/CMakeLists.txt:
| Imported target "nvbench_json" includes non-existent path

| "/include"
| in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

| * The path was deleted, renamed, or moved to another location.

| * An install or uninstall procedure did not complete successfully.

| * The installation package was faulty and references files it does not
| provide.

...

meson.build:10:0: ERROR: Failed to configure the CMake subproject
```

The location in question appears to be adding `"${nlohmann_json_SOURCE_DIR}/include"` but without that actually existing. Also that's kind of annoying to do at all, because it should be looking up nlohmann_json's official installable `*-config.cmake` files via find_package(), but instead nvbench is downloading https://github.com/nlohmann/json/releases/download/v3.9.1/include.zip on the fly and copying files out of there.

**Comment by bsarden** [comment]
Strange, it looks like the NVBench authors are doing that so that they can [patch nlohmann/json to work with nvcc](https://github.com/NVIDIA/nvbench/commit/9cfb15be8b4d2c7b679857cc3e58c1ab2ed3fc1c). How should we make forward progress here?

**Comment by eli-schwartz** [comment]
Not only did the nvbench authors use a locally downloaded copy of nlohmann-json long before adding that patch, but also it seems they never reported the bug to the nlohmann-json project.

Sad.

...

Anyway the problem *here* is that I simply don't understand cmake well enough to guess why it errors out here, so as a result I can't progress far enough to duplicate your issue and therefore have no clue what to even try fixing.

But maybe @mensinda has more ideas.

**Comment by mensinda** [comment]
OK, found the problem: There is a CMake custom target `nvbench.all`:

```cmake
# Builds all NVBench targets (libs, tests, examples, etc).
add_custom_target(nvbench.all)
```

This target depends on `nvbench` and because there exists a `nvbench` directory, the `'nvbench'` string is wrongly added to the inputs of the `nvbench_all` Meson `custom_target`.

Should be fixed with #10303.
